sdk/go/src/sawtooth_sdk/client/encoder.go

```diff
@@ -0,0 +1,270 @@
+package client
+
+import (
+	"encoding/hex"
+	"fmt"
+	"github.com/golang/protobuf/proto"
+	"sawtooth_sdk/protobuf/batch_pb2"
+	"sawtooth_sdk/protobuf/transaction_pb2"
+	"strings"
+	"time"
+)
+
+type TransactionParams struct {
+	FamilyName      string
+	FamilyVersion   string
+	PayloadEncoding string
+	Nonce           string
+	BatcherPubkey   string
+	Dependencies    []string
+	Inputs          []string
+	Outputs         []string
+}
+
+type Encoder struct {
+	privkey  []byte
+	pubkey   string
+	defaults TransactionParams
+}
+
+// NewTransactionEncoder constructs a new encoder which can be used to generate
+// transactions and batches, and to serialize batches for submitting to the
+// REST API.
+func NewEncoder(privkey []byte, defaults TransactionParams) *Encoder {
+	return &Encoder{
+		privkey:  privkey,
+		pubkey:   Encode(GenPubKey(privkey), "hex"),
+		defaults: defaults,
+	}
+}
+
+// -- Transactions --
+
+// NewTransaction Creates a new transaction and handles the construction and
+// signing of the transaction header.
+func (self *Encoder) NewTransaction(payload []byte, p TransactionParams) *Transaction {
+	h := &transaction_pb2.TransactionHeader{
+		// Load defaults
+		FamilyName:      self.defaults.FamilyName,
+		FamilyVersion:   self.defaults.FamilyVersion,
+		PayloadEncoding: self.defaults.PayloadEncoding,
+		Nonce:           self.defaults.Nonce,
```

Comment:
I see how this streamlines the encoder interface . . . but it would be awfully weird to have a default nonce.